### PR TITLE
Split (unit)tests and add running OS support

### LINKED ISSUE #682: Support (unit) tests that may need a running OS
**What would you like to be added**:
We need support for (unit) tests that may need to test and execute commands within a running OS.
This means that we need to run the final artifact in a VM and run the tests there. Therefore, we need to adjust the test handling to support this.

**Why is this needed**:
With #647 (CIS integration) and #648 (FedRAMP integration) we need to execute these scripts within a running OS (not only a chroot).

### LINKED ISSUE #683: Split (unit) tests from garden-build
**What would you like to be added**:
Currently the whole logic for (unit) tests is built into *garden-build* which makes it very complex and difficult to handle tests that should be called from a final artifact. Therefore, we should split this and create a new *garden-test* script that contains all the needed tasks. Next to this, this should be called after the artifact/image got created and perform the tests. 

**Why is this needed**:
We will need further tests that may need a final artifact, as well as a executing tests on a running OS.

### PR DESCRIPTION
**Pull request by gyptazy**
This PR contains three changes for:
 * #683: Split (unit) tests from garden-build
 * #682: Support (unit) tests that may need a running OS
 * #647: Example integration for upcoming features (example: cis)

**How to categorize this PR?**

/kind TODO
/area os
/os garden-linux

**What this PR does / why we need it**:
All changes are needed to solve #648 and #647

**Which issue(s) this PR fixes**:
Fixes #683 
Fixes #682 
Relates #647

**Special notes for your reviewer**:

**Release note**:
Still in draft!

- Creates a new *garden-test* script
- *garden-test* will be executed after the final artifact is created
- Takes the rootfs from artifact and modifies it for special use cases (e.g. tests that may need a running OS)
- Watches for *.startvm* files in features; these will be used as a trigger
- For each feature with a *startvm* script, the rootfs will be modified with the needed content and recreated as a image
- Within the docker instance the image will be booted via QEMU
- Remote SSH will perform and validate the executed (unit) tests

ToDo:
- Since I was unable to use the already present *start-vm* script, I directly call *qemu-system-x86_64*. Therefore, I still need to integrate an $arch switch.
- Too many static items
- [...]

```feature user

```

### CODE REVIEW

**Comment by gyptazy** [comment]
Ready for review.

File: bin/garden-test [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+		# Test: startvm (running in qemu)
+                elif [ ${t##*.} == "startvm" ]; then
+                        cp -L ${actualTest} "${rootfs}/tmp/${test}"
+			# Strip trailing whitespace
+			startvmUser=$(echo ${enabledBy} | head -c-1)
+			# Create a temporary and only for unit tests valid
+			# ssh key for each user and feature
+			garden-chroot "${rootfs}" useradd -m ${startvmUser}
+			garden-chroot "${rootfs}" adduser ${startvmUser} wheel
```
Comment (nanory) [inline bin/garden-test:208]: Please use this instead:
```
useradd -m -G wheel ${startvmUser}
```

File: bin/garden-test [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+			garden-chroot "${rootfs}" mkdir -p /home/${startvmUser}/tests/
+			garden-chroot "${rootfs}" mkdir -p /home/${startvmUser}/.ssh/
+			garden-chroot "${rootfs}" chmod 700 /home/${startvmUser}/.ssh/
+			/usr/bin/ssh-keygen -q -t rsa -N '' -f /rootfs/home/${startvmUser}/.ssh/id_rsa -m pem
+			cp /rootfs/home/${startvmUser}/.ssh/id_rsa.pub /rootfs/home/${startvmUser}/.ssh/authorized_keys
+			garden-chroot "${rootfs}" chmod 600 /home/${startvmUser}/.ssh/authorized_keys
+			garden-chroot "${rootfs}" chown -R ${startvmUser}:${startvmUser} /home/${startvmUser}/.ssh/
```
Comment (nanory) [inline bin/garden-test:215]: All of these commands do not require to execute them via `garden-chroot`. You can also use the test-image context for this.

File: bin/garden-test [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                        cp -L ${actualTest} "${rootfs}/tmp/${test}"
+			# Strip trailing whitespace
+			startvmUser=$(echo ${enabledBy} | head -c-1)
+			# Create a temporary and only for unit tests valid
+			# ssh key for each user and feature
+			garden-chroot "${rootfs}" useradd -m ${startvmUser}
+			garden-chroot "${rootfs}" adduser ${startvmUser} wheel
+			garden-chroot "${rootfs}" mkdir -p /home/${startvmUser}/tests/
+			garden-chroot "${rootfs}" mkdir -p /home/${startvmUser}/.ssh/
+			garden-chroot "${rootfs}" chmod 700 /home/${startvmUser}/.ssh/
```
Comment (nanory) [inline bin/garden-test:211]: Please use this instead:
```
mkdir -p ${rootfs}/home/${startvmUser}/.ssh -m 0700
```

File: bin/garden-test [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+			/usr/bin/ssh-keygen -q -t rsa -N '' -f /rootfs/home/${startvmUser}/.ssh/id_rsa -m pem
+			cp /rootfs/home/${startvmUser}/.ssh/id_rsa.pub /rootfs/home/${startvmUser}/.ssh/authorized_keys
+			garden-chroot "${rootfs}" chmod 600 /home/${startvmUser}/.ssh/authorized_keys
+			garden-chroot "${rootfs}" chown -R ${startvmUser}:${startvmUser} /home/${startvmUser}/.ssh/
+			# Copy whole test directory to features home directory
+			rsync -aP "${featureDir}/${startvmUser}/test/" /rootfs/home/${startvmUser}/tests/
```
Comment (nanory) [inline bin/garden-test:217]: Why do we copy the tests into the rootfs again, if they are already copied into the rootfs some lines above?
The tests will be underneath the `/tmp/<test>/<test>.d` directory already.

File: bin/garden-test [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                                echo "Starting VM for ${arch}"
+                                qemu-system-aarch64 -display none -daemonize -cpu cortex-a72 -machine virt -pidfile /tmp/qemu.pid -bios /usr/share/qemu-efi-aarch64/QEMU_EFI.fd -m 1024M -device virtio-net-pci,netdev=net0,mac=02:9f:ec:22:f8:89 -netdev user,id=net0,hostfwd=tcp::2223-:22,hostname=garden /tmp/test.raw
+                        fi
+
+			# Wait for SSH in VM to be present
+			# Default: tcp/2223
+			echo "Starting VM - this will take some time"
+			sshwatch 2223
+			# Finally perform test
+			ssh -i /rootfs/home/${startvmUser}/.ssh/id_rsa ${startvmUser}@localhost -p 2223 -o StrictHostKeyChecking=no /home/${startvmUser}/tests/${test}.startvm
```
Comment (nanory) [inline bin/garden-test:245]: Can we split this line into multiple ones?

File: features/cis/exec.config [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+set -Eeuo pipefail
+
+echo "dummy cis run"
+exit 0
```
Comment (nanory) [inline features/cis/exec.config:5]: Is this file really needed right now?

**Review by nanory — CHANGES_REQUESTED** [review]
Hi @gyptazy, I have added some comments to this MR. If something is unclear, we can have a small discussion.

File: bin/garden-test (hunk 1 above)
Comment (gyptazy) [reply bin/garden-test:208]: Ok, but still within the chroot.

File: bin/garden-test (hunk 2 above)
Comment (gyptazy) [reply bin/garden-test:215]: Basically it does. *$startvmUser* resolves to a user that is only known within the chroot (created within l207). Changing the user to a not known one will result in a failure. However, I think it's still a good indicator to clarify that this commands are executed in a chroot instead the test-image.

File: bin/garden-test (hunk 4 above)
Comment (gyptazy) [reply bin/garden-test:217]: Yep, it's */tmp* and the content isn't present anymore when the VM boots up in QEMU. Maybe we should move the overall tests to a new persistent directory since the new test image is dropped in any case. But this should be done in a new PR.

File: bin/garden-test (hunk 1 above)
Comment (nanory) [reply bin/garden-test:208]: Yes, sound good.

File: bin/garden-test [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+                        	fi
+                	done
+			# Start new generated image for unit tests
+                        if [ "$arch" = amd64 ]; then
+                                echo "Starting VM for ${arch}"
+                                qemu-system-x86_64 -display none -daemonize -pidfile /tmp/qemu.pid -m 1024M -device virtio-net-pci,netdev=net0,mac=02:9f:ec:22:f8:89 -netdev user,id=net0,hostfwd=tcp::2223-:22,hostname=garden /tmp/test.raw
+                        elif [ "$arch" = arm64 ]; then
+                                echo "Starting VM for ${arch}"
+                                qemu-system-aarch64 -display none -daemonize -cpu cortex-a72 -machine virt -pidfile /tmp/qemu.pid -bios /usr/share/qemu-efi-aarch64/QEMU_EFI.fd -m 1024M -device virtio-net-pci,netdev=net0,mac=02:9f:ec:22:f8:89 -netdev user,id=net0,hostfwd=tcp::2223-:22,hostname=garden /tmp/test.raw
+                        fi
```
[the remark this answers is not in the document]
Comment (gyptazy) [reply bin/garden-test:245]: I think we may create another issue for this to make *start-vm* compatible for this use case. For the first iteration we should go with this implementation.

File: bin/garden-test [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+mkdir -p rootfs/etc/ssl
+mount --bind /etc/ssl rootfs/etc/ssl
+garden-apt-get $rootfs update
+
+# Define APT settings and install
+# needed dependencies
+sed -i 's/testing/bookworm/g' /etc/apt/sources.list
+rm /etc/apt/apt.conf.d/default-testing
+apt-get update
+apt-get -y install procps iproute2 rsync openssh-client qemu-system-arm qemu-system-x86
```
[the remark this answers is not in the document]
Comment (gyptazy) [reply bin/garden-test]: Ok, this speeds it up. But instead of creating a new image I decided to integrate these few packages into the build image.

**Review by nanory — APPROVED** [review]

**Review by gehoern — APPROVED** [review]
